noria-server/mir/src/optimize.rs

```diff
@@ -1,19 +1,217 @@
+        }
+
+        visited_nodes.insert(node_name, true);
+        found_nodes.push(n);
+    }
+
+    // 2. iterate over the nodes to find candidates, i.e.
+    // nodes that are a filter followed by an aggregate with no other children/parents.
+    // TODO (jamb) Initially we also considered nodes that are an aggregate followed by a filter,
+    // but this case is rarer and may have the problem that it is filtering on the result
+    // of the aggregate, in which case they cannot be merged.
+
+    let mut candidate_nodes = Vec::new();
+
+    for n in found_nodes {
+        // if we've already set this to false, we shouldn't look at it more;
+        // e.g. it might be merging with its parent node.
+        let node_name = n.borrow().versioned_name();
+        if !visited_nodes.get(&node_name).unwrap() {
+            continue;
+        }
+        // now scan for candidacy
+        let mut candidate = false;
+        match n.borrow().inner {
+            MirNodeType::Filter { .. }  => {
+                // if there's exactly one child and it's an aggregation and it has exactly one parent,
+                // then this is a candidate
+                if n.borrow().children.len() == 1 {
+                    let temp = n.borrow();
+                    let child = temp.children.first().unwrap().borrow();
+                    match child.inner {
+                        MirNodeType::Aggregation { .. } => {
+                            if child.ancestors.len() == 1 {
+                                candidate = true;
+                            }
+                        },
+                        _ => {},
+                    };
+                }
+            },
+            MirNodeType::Aggregation { ref on, .. }  => {
+                // if there's exactly one child and it's an aggregation and it has exactly one parent,
+                // then this is a candidate
+                if n.borrow().children.len() == 1 {
+                    let temp = n.borrow();
+                    let child = temp.children.first().unwrap().borrow();
+                    match child.inner {
+                        MirNodeType::Filter { ref conditions } => {
+                            if child.ancestors.len() == 1 {
+                                candidate = true;
+
+                                // But wait -- need to check if the filter is on the aggregation result
+                                use nom_sql::FunctionExpression::{Count, Sum};
+                                for (i, col) in child.columns.iter().enumerate() {
+                                    match col.function.clone() {
+                                        Some(func_expr) => {
+                                            match *func_expr {
+                                                 Count(col, _)
+                                                 | Sum(col, _) => {
+                                                    if col.name == on.name {
```

Comment:
This can be a match guard instead (just move it before the `=>`) to save another level of indentation.

Reply:
Same here. This looks a lot less horrendously over-nested now :)